### PR TITLE
Add additional metadata to the columns, dataset and benchmark

### LINKED ISSUE #18: Feat: Add precomputed fields
We need pydantic 2.x to be able to do that without an horrible hack :-) No rush and blocked by https://github.com/conda-forge/pydantic-feedstock/pull/82

The idea is to add a few pre-computed fields whenever it makes sense.

A few examples for `Dataset`:

```python
  repr_dict["n_datapoints"] = len(self)
  repr_dict["n_columns"] = len(self.table.columns)
  repr_dict["dataset_size_mb"] = 128
```

### LINKED ISSUE #63: Additional attributes for Benchmark
# Description
To facilitate for both UI searching and downstream usage (such as identify the metrics for the corresponding scenario), 
I suggest adding attributes for Benchmark 
`task_type`: reg or cls
`num_classes`: for multiclass tasks , default binary

# Acceptance Criteria
- Attributes have been added.

### PR DESCRIPTION
**Pull request by cwognum**
## Changelogs

**Metadata for benchmarks**
- Fixes #63 
  - Adds the task type (i.e. multi task or single task), computed automatically. 
  - Adds the target type (i.e. regression or classification), can be manually specified, but if not lib tries to infer automatically using [`sklearn`'s `type_of_target()`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.multiclass.type_of_target.html).
  - Adds the train set size, computed automatically. 
  - Adds the test set(s) size(s), computed automatically. 
  - Adds the number of classes for classification tasks, computed automatically. 
  - Adds the number of test sets

**Metadata for datasets**
- Fixes #18
  - Adds the dtype to the column annotations, computed automatically. 
  - Adds the no. datapoints and no. columns, computed automatically. 
  - Adds a n optional field with a reference to the curation. 

**Misc**
- Add additional syntax to the dataset to make interacting with it easier, similar to `pandas.DataFrame`. E.g. `dataset[row, col]` or `dataset[:, "smiles"]`. 

---

_Checklist:_

- [X] _Was this PR discussed in an issue? It is recommended to first discuss a new feature into a GitHub issue before opening a PR._
- [X] _Add tests to cover the fixed bug(s) or the newly introduced feature(s) (if appropriate)._
- [X] _Update the API documentation if a new function is added, or an existing one is deleted._
- [X] _Write concise and explanatory changelogs above._
- [X] _If possible, assign one of the following labels to the PR: `feature`, `fix` or `test` (or ask a maintainer to do it for you)._

### CODE REVIEW

**Comment by cwognum** [comment]
Worth noting that we will also have the size of the dataset (in bytes), but I will save this on the Hub side since the Hub already receives this information anyways.

**Review by hadim — APPROVED** [review]

**Comment by cwognum** [comment]
Quick! Change from `meta-data` to `metadata` before @jstlaurent and @hadim spot it! :eyes:

File: polaris/benchmark/_base.py [hunk 1]
```diff
@@ -254,6 +309,47 @@ def _compute_checksum(dataset, target_cols, input_cols, split, metrics):
...
+    @property
+    def n_classes(self) -> dict[str, int]:
+        """The number of classes for each of the target columns."""
+        n_classes = {}
+        for target in self.target_cols:
+            target_type = self.target_types[target]
+            if target_type is None or target_type == TargetType.REGRESSION:
+                n_classes[target] = None
+            else:
+                n_classes[target] = self.dataset.loc[:, target].nunique()
```
[the remark this answers is not in the document]
Comment (cwognum) [reply polaris/benchmark/_base.py:343]: Good point, but I don't think so! `nunique()` excludes NaNs by default: https://pandas.pydata.org/docs/reference/api/pandas.Series.nunique.html

**Review by zhu0619 — APPROVED** [review]
